master/internal/api_runs.go

```diff
@@ -403,3 +420,90 @@ func (a *apiServer) MoveRuns(
 	}
 	return &apiv1.MoveRunsResponse{Results: results}, nil
 }
+
+func (a *apiServer) KillRuns(ctx context.Context, req *apiv1.KillRunsRequest,
+) (*apiv1.KillRunsResponse, error) {
+	curUser, _, err := grpcutil.GetUser(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	var runChecks []archiveKillRunOKResult
+	getQ := getSelectRunsQueryTables().
+		Model(&runChecks).
+		Join("LEFT JOIN trials_v2 t ON r.id=t.run_id").
+		Column("r.id").
+		ColumnExpr("COALESCE((e.archived OR p.archived OR w.archived), FALSE) AS archived").
+		ColumnExpr("t.request_id").
+		ColumnExpr("r.state IN (?) AS state", bun.In(model.StatesToStrings(model.TerminalStates))).
+		Where("r.project_id = ?", req.ProjectId)
+
+	if req.Filter == nil {
+		getQ = getQ.Where("r.id IN (?)", bun.In(req.RunIds))
+	} else {
+		getQ, err = filterRunQuery(getQ, req.Filter)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if getQ, err = experiment.AuthZProvider.Get().
+		FilterExperimentsQuery(ctx, *curUser, nil, getQ,
+			[]rbacv1.PermissionType{rbacv1.PermissionType_PERMISSION_TYPE_UPDATE_EXPERIMENT}); err != nil {
+		return nil, err
+	}
+
+	err = getQ.Scan(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	var results []*apiv1.RunActionResult
+	var visibleIDs []int32
+	var validIDs []int32
+	for _, check := range runChecks {
+		visibleIDs = append(visibleIDs, check.ID)
+		switch {
+		case check.State:
+			results = append(results, &apiv1.RunActionResult{
+				Error: "",
+				Id:    check.ID,
+			})
+		case check.RequestID == nil:
+			results = append(results, &apiv1.RunActionResult{
+				Error: "Run has no associated request id.",
+				Id:    check.ID,
+			})
+		default:
+			validIDs = append(validIDs, check.ID)
+		}
+	}
+	if req.Filter == nil {
+		for _, originalID := range req.RunIds {
+			if !slices.Contains(visibleIDs, originalID) {
+				results = append(results, &apiv1.RunActionResult{
+					Error: fmt.Sprintf("Run with id '%d' not found", originalID),
+					Id:    originalID,
+				})
+			}
+		}
+	}
+
+	for _, runID := range validIDs {
+		_, err = a.KillTrial(ctx, &apiv1.KillTrialRequest{
+			Id: runID,
+		})
+		if err != nil {
+			results = append(results, &apiv1.RunActionResult{
```

Comment:
i guess after really sinking into this PR, i've realized this "wanting a result back for everything" really complicates it. how do we plan to use this result? how necessary is it?

because I was thinking, why are we not just doing base select with runs, filtering for killable runs by adding a `WHERE state in (...terminal states) and request_id IS NOT NULL`, then passing those all to kill trial. at that point, if there are any errors we can pass them back to the user? that seems sufficient and twice as simple?

Reply:
This is how the responses were structured in the Experiment equivalent functions (PauseExperiments, DeleteExperiments, etc...) I believe as well that Frontend expects this behaviour.